### PR TITLE
chore(docs): miscellaneous fixes to generate config schema and derived docs output

### PR DESCRIPTION
**Pull request by tobz**
This PR fixes two small glitches:
- adding (un)signed integer fields to the configuration schema as `number` (which implies floating-point number) instead of `integer`
- generating the incorrect numeric type for numeric values that are part of any "enum" values in a schema

The first thing is simple: JSON Schema supports two numeric type, `integer` and `number`, which represent whole integers (`u32`, `i64`, etc etc) and floating-point numbers, respectively. We were emitting `number` in all cases, even though we had access to the necessary data to properly select either `integer` or `number`. We fixed that with a mere few changed lines.

The second thing, and the thing that represents like 95% of this PR, is fixing the numeric type used to describe numeric values in an "enum" schema.

Essentially, in the `CompressionLevel` type, we allow some fixed strings (good, best, fast, etc) and a fixed set of levels: 0-9, inclusive. When generated as part of the configuration schema, we simply emit those values in a single JSON array. When machine generating documentation from the configuration schema, we would emit something that looks like:

```
type:
  string: enum: ["good", "fast", "best"],
  integer: enum :[0, 1, 3, ..., 9],
```

Normally in the configuration schema, for fields with numeric types, we add a metadata annotation that describes the actual numeric type -- `uint`, `int`, or `float` -- to provide clarity where JSON's own types cannot. Instead of `integer` or `number`, we actually use one of those three, more precise values.

In the case of enum values, though, we can't do that because it wouldn't be clear what that annotation applies to amongst all the different enum values. So, the meat of this PR: we now use a simple heuristic when grouping enum values by their type in order to promote `integer`/`number` to something more specific if we can.

The logic is simple and is centered around seeing if all the values would fit in an `int` or a `uint` or a `float` -- which represent `i64`, `u64`, and `f64`, respectively -- and basically just picks the type that is most constrained, with a preference for `uint`, because most of our numeric fields actually are unsigned. It's not perfect, and it's a simple heuristic, but it works and is at least more descriptive than what was being generated before.

Since, as you might surmise, such logic would only take tens of lines, not hundreds, you're wondering: what the hell is the rest of the diff, then? Simply put: all the related changes to thread the "source" schema -- the unfettered configuration schema -- through the necessary places so that we can try and look up the numeric type metadata annotation where possible for maximum precision, rather than always falling back to the heuristic.

It's cursed code, yes, and I'm going to eventually throw it all away and rewrite this in Rust... but until then, you can take solace in the isolated diff of the machine-generated output itself, and that I will never foist the maintenance of this thing, in its Ruby form, upon any of you.

### CODE REVIEW

File: lib/vector-config/src/schema.rs [hunk 1]
```diff
@@ -151,13 +151,16 @@ pub fn generate_number_schema<N>() -> SchemaObject
 where
     N: Configurable + ConfigurableNumber,
 {
+    // TODO: Once `schemars` has proper integer support, we should allow specifying min/max bounds
```
Comment (spencergilbert) [inline lib/vector-config/src/schema.rs:154]: Is there a tracking issue or similar we can link here?

File: scripts/generate-component-docs.rb [hunk 2]
```diff
@@ -300,21 +462,121 @@ def apply_object_property_fields!(parent_schema, property_schema, property_name,
...
+# If the schema contains a top-level schema reference, or if any of the parts of its schema contain
+# schema references (array items schema, any subschemas in `oneOf`/`allOf`, etc), then those
+# references are expanded. Expansion happens recursively until all schema references
 #
 # For any overlapping fields in the given schema and the referenced schema, the fields from the
 # given schema will win.
-def resolve_schema_reference(root, schema)
-  schema = deep_copy(schema)
+def expand_schema_references(root_schema, unexpanded_schema)
+  # Break any cycles during expansion, to same as we do during resolution.
```
Comment (spencergilbert) [inline scripts/generate-component-docs.rb:474]:
```suggestion
  # Break any cycles during expansion, the same as we do during resolution.
```

File: scripts/generate-component-docs.rb [hunk 3]
```diff
@@ -375,8 +649,10 @@ def get_reduced_resolved_schema(schema)
 
 # Fully resolves a schema definition, if it exists.
 #
-# This looks up a schema definition by the given `name` within `root_schema` and resolves it, if it
-# exists, Otherwise, `nil` is returned.
+# This looks up a schema definition by the given `name` within `root_schema` and resolves i
```
Comment (spencergilbert) [inline scripts/generate-component-docs.rb:652]:
```suggestion
# This looks up a schema definition by the given `name` within `root_schema` and resolves it.
```

**Review by spencergilbert — APPROVED** [review]
That ruby is definitely painful

File: lib/vector-config/src/schema.rs (hunk 1 above)
Comment (tobz) [reply lib/vector-config/src/schema.rs:154]: I'll add an issue for that, good call.

Comment (tobz) [reply lib/vector-config/src/schema.rs:154]: Added: https://github.com/vectordotdev/vector/issues/15271

**Pull request merged**